libdevcrypto/LibSnark.cpp

```diff
@@ -0,0 +1,213 @@
+	if (!p.is_well_formed())
+		BOOST_THROW_EXCEPTION(InvalidEncoding());
+	return p;
+}
+
+bytes encodePointG1(libff::alt_bn128_G1 _p)
+{
+	if (_p.is_zero())
+		return bytes(64, 0);
+	_p.to_affine_coordinates();
+	return
+		fromLibsnarkBigint(_p.X.as_bigint()).asBytes() +
+		fromLibsnarkBigint(_p.Y.as_bigint()).asBytes();
+}
+
+libff::alt_bn128_Fq2 decodeFq2Element(dev::bytesConstRef _data)
+{
+	// Encoding: c1 (256 bits) c0 (256 bits)
+	// "Big endian", just like the numbers
+	return libff::alt_bn128_Fq2(
+		decodeFqElement(_data.cropped(32)),
+		decodeFqElement(_data.cropped(0))
+	);
+}
+
+libff::alt_bn128_G2 decodePointG2(dev::bytesConstRef _data)
+{
+	libff::alt_bn128_Fq2 x = decodeFq2Element(_data);
+	libff::alt_bn128_Fq2 y = decodeFq2Element(_data.cropped(64));
+	if (x == libff::alt_bn128_Fq2::zero() && y == libff::alt_bn128_Fq2::zero())
+		return libff::alt_bn128_G2::zero();
+	libff::alt_bn128_G2 p(x, y, libff::alt_bn128_Fq2::one());
+	if (!p.is_well_formed())
+		BOOST_THROW_EXCEPTION(InvalidEncoding());
+	return p;
+}
+
+}
+
+pair<bool, bytes> dev::crypto::alt_bn128_pairing_product(dev::bytesConstRef _in)
+{
+	// Input: list of pairs of G1 and G2 points
+	// Output: 1 if pairing evaluates to 1, 0 otherwise (left-padded to 32 bytes)
+
+	size_t constexpr pairSize = 2 * 32 + 2 * 64;
+	size_t const pairs = _in.size() / pairSize;
+	if (pairs * pairSize != _in.size())
+	{
+		// Invalid length.
+		return {false, bytes{}};
+	}
+	try
+	{
+		initLibSnark();
+		libff::alt_bn128_Fq12 x = libff::alt_bn128_Fq12::one();
+		for (size_t i = 0; i < pairs; ++i)
+		{
+			dev::bytesConstRef pair = _in.cropped(i * pairSize, pairSize);
+			libff::alt_bn128_G2 p = decodePointG2(pair.cropped(2 * 32));
+			if (-libff::alt_bn128_G2::scalar_field::one() * p + p != libff::alt_bn128_G2::zero())
+				// p is not an element of the group (has wrong order)
+				return {false, bytes()};
+			if (p.is_zero())
+				continue; // the pairing is one
+			if (decodePointG1(pair).is_zero())
+				continue; // the pairing is one
+			x = x * libff::alt_bn128_miller_loop(
+				libff::alt_bn128_precompute_G1(decodePointG1(pair)),
+				libff::alt_bn128_precompute_G2(p)
+			);
+		}
+		bool result = libff::alt_bn128_final_exponentiation(x) == libff::alt_bn128_GT::one();
+		return {true, h256{result}.asBytes()};
+	}
+	catch (InvalidEncoding const&)
+	{
+		return {false, bytes{}};
+	}
+	catch (...)
+	{
+		cwarn << "Internal exception from libsnark. Forwarding as precompiled contract failure.";
+		return {false, bytes{}};
+	}
+}
+
+pair<bool, bytes> dev::crypto::alt_bn128_G1_add(dev::bytesConstRef _in)
+{
+	try
+	{
+		initLibSnark();
+		libff::alt_bn128_G1 const p1 = decodePointG1(_in);
+		libff::alt_bn128_G1 const p2 = decodePointG1(_in.cropped(32 * 2));
+		return {true, encodePointG1(p1 + p2)};
+	}
+	catch (InvalidEncoding const&)
+	{
+	}
+	catch (...)
+	{
+		cwarn << "Internal exception from libsnark. Forwarding as precompiled contract failure.";
```

Comment:
I prefer not to. It's better to terminate the node in this case than running it in invalid state mode. Agree?